### PR TITLE
Feature/aasxgeneration

### PR DESCRIPTION
**Pull request by bs-jokri**
This PR contributes a AAS generator module to the SDS-SDK. It allows to map BAMM Aspect Models to Asset Administration Shell submodel templates. The module is also usable via a BAMM CLI command.

### CODE REVIEW

File: core/sds-aspect-model-aas-generator/src/main/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASVisitor.java [hunk 1]
```diff
@@ -0,0 +1,581 @@
...
+               .values( Collections.singletonList( decideOnMapping( property, context ) ) )
+               .ordered( true )
+               .allowDuplicates( false )
+               .build();
+      createSubmodelElement( builder, context );
+      return context.getEnvironment();
+   }
+
+   // Either will be mapped by adding both the left and the right side to the SubmodelTemplate.
+   // TODO check whether there are possibilities to mark the SubmodelElements as optional
```
Comment (atextor) [inline core/sds-aspect-model-aas-generator/src/main/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASVisitor.java:477]: Clarify and remove TODO

File: core/sds-aspect-model-aas-generator/src/main/java/io/openmanufacturing/sds/aspectmodel/aas/Context.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      this.submodel = ofInterest;
+   }
+
+   public boolean hasEnvironmentConceptDescription( String id ) {
+      return getEnvironment().getConceptDescriptions().stream().anyMatch( x -> x.getIdentification().getIdentifier().equals(id) );
+   }
+
+   public ConceptDescription getConceptDescription( String id ) {
+      Optional<ConceptDescription> optional = getEnvironment().getConceptDescriptions().stream().filter( x -> x.getIdentification().getIdentifier().equals( id ) ).findFirst();
+      if(optional.isEmpty()){
```
Comment (atextor) [inline core/sds-aspect-model-aas-generator/src/main/java/io/openmanufacturing/sds/aspectmodel/aas/Context.java:42]: Please run code formatter

File: core/sds-aspect-model-aas-generator/src/test/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASGeneratorTest.java [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+
+      assertEquals( 1, env.getSubmodels().size() );
+      assertEquals( 1, env.getSubmodels().get( 0 ).getSubmodelElements().size() );
+      Property submodelElement = (Property) env.getSubmodels().get( 0 ).getSubmodelElements().get( 0 );
+      assertEquals( "http://www.w3.org/2001/XMLSchema#integer", submodelElement.getValueType(), "Value type not int" );
+   }
+
+   @ParameterizedTest
+   @EnumSource( value = TestAspect.class, mode = EnumSource.Mode.EXCLUDE, names = {
+         "ASPECT_WITH_ABSTRACT_ENTITY", // Aspect model not present
```
Comment (atextor) [inline core/sds-aspect-model-aas-generator/src/test/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASGeneratorTest.java:222]: These Aspect models are not found because the `TestAspect` enum refers to the set of test models that are present for the latest meta model version (the latest and greatest set). This fails here because the `loadAspect()` method has loading test models for BAMM 1.0.0 hard coded:
`TestResources.getModel( testAspect, KnownVersion.BAMM_1_0_0 ).get()`
This should be replace by the following:
`TestResources.getModel( testAspect, KnownVersion.getLatest() ).get()`

Then the excludes here can be removed!

File: documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc [hunk 4]
```diff
@@ -439,3 +439,82 @@ characteristicDefinitionsUrl.ifPresent( url -> {
...
+The Asset Administration Shell (AAS) and its https://www.plattform-i40.de/IP/Redaktion/EN/Standardartikel/specification-administrationshell.html[information model] is a widely recognized standard developed by the https://industrialdigitaltwin.org[Industrial Digital Twin Association (IDTA)] to express and handle Digital Twins. Central element of the AAS is
+the concept of Submodels, which describe certain aspects of a Digital Twin.
+
+The BAMM Aspect Meta Model allows to specify aspects of a digital twin and its semantics.
+The AAS Generator module provides mapping implementations to derive AAS Submodels from BAMM Aspect models
+and by that allows to on the one hand integrate BAMM models in AAS environments and on the other hand allow
+AAS Submodels to be described with rich semantics, as it is possible with BAMM.
+
+[[details-mapping-aas]]
+== Details to the Mapping Concept
```
Comment (atextor) [inline documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc:453]: This should be one indentation level below, i.e. "===". Also, "Details _of_ the Mapping Concept"?

File: documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc [hunk 5]
```diff
@@ -439,3 +439,82 @@ characteristicDefinitionsUrl.ifPresent( url -> {
...
 ----
 
+[[mapping-aas]]
+== Mapping Aspect Models to Asset Administration Shell (AAS) Submodel Templates
+The Asset Administration Shell (AAS) and its https://www.plattform-i40.de/IP/Redaktion/EN/Standardartikel/specification-administrationshell.html[information model] is a widely recognized standard developed by the https://industrialdigitaltwin.org[Industrial Digital Twin Association (IDTA)] to express and handle Digital Twins. Central element of the AAS is
+the concept of Submodels, which describe certain aspects of a Digital Twin.
+
+The BAMM Aspect Meta Model allows to specify aspects of a digital twin and its semantics.
+The AAS Generator module provides mapping implementations to derive AAS Submodels from BAMM Aspect models
+and by that allows to on the one hand integrate BAMM models in AAS environments and on the other hand allow
```
Comment (atextor) [inline documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc:449]: "to on the one hand..." sounds a bit strange. Could we rewrite this sentence like this?:
> The AAS Generator module provides mapping implementations to derive AAS Submodels from BAMM Aspect Models. On the one hand, this allows to integrate BAMM models in AAS environments and on the other hand it allows AAS submodels to be described with rich semantics, as it is possible with BAMM.

File: documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc [hunk 6]
```diff
@@ -439,3 +439,82 @@ characteristicDefinitionsUrl.ifPresent( url -> {
...
+
+| bamm:description   | aas:Submodel.description  |
+
+| bamm:property   | see **bamm:Property**  |
+
+| bamm:operation   | see **bamm:Operation**   |
+
+| bamm:Aspect.urn  | aas:Submodel.semanticId  |
+
+| **Property**  | aas:Property <br> aas:SubmodelElementCollection   | The AAS type is derived from the type of the BAMM Characteristic specifying the BAMM property. Depending on the type it is decided what the resulting AAS element will be. In case of an Entity it will result in a SubmodelElementCollection. It will also be a SubmodelElementCollection if the BAMM Characteristic is of a Collection type (see the https://openmanufacturingplatform.github.io/sds-documentation/bamm-specification/v1.0.0/characteristics.html[Characteristics taxonomy]). In all other cases an aas:Property will be generated
```
Comment (atextor) [inline documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc:476]: Literal `<br>` will not work in Antora. To force a line break inside the cell, allow Asciidoc-Markup inside the table cell by marking the column `a`, i.e. change the line above the table to `[cols="1,1a,2"]`, then in the cell, instead of `<br>`, add a single empty line. Repeat for other `<br>` occurences in the table.

Also please check whether it is correct that there is no comma between the referenced elements in the second colum, while other cells below mention two elements but separate them using a comma.

Comment (atextor) [inline documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc:476]: Replace the hyperlink with a proper `xref`, i.e. `xref:bamm-specification:ROOT:characteristics.adoc[Characteristics taxonomy]`

File: documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc [hunk 7]
```diff
@@ -439,3 +439,82 @@ characteristicDefinitionsUrl.ifPresent( url -> {
...
+
+| **Operation** |  Operation  | in/out parameters are not used in BAMM so the mapping only generates input variables and output variables in AAS
+
+| **Characteristic**  | aas:SubmodelElement, aas:ConceptDescription  | Characteristics in BAMM define the semantics of a property, which includes there types as well as links to further definitions (standards, dictionaries, etc), a natural language description and name in different languages. Type and description are separated in AAS, which is why there is not a one-to-one mapping of a Characteristic to one element in AAS but rather Characteristics are used in the mapping of Properties, first, to guide the generation process and, second, to capture semantics in ConceptDescriptions of properties with data specification "DataSpecificationIEC61360" of the AAS.
+
+| **Collection**  | aas:SubmodelElementCollection, aas:ConceptDescription  | The general remarks to Characteristics apply also for Collection type Characteristics. However, properties referencing Collections are mapped to SubmodelElementCollections. Specific properties of collections are mapped. bamm:Set is unique, bamm:SortedSet is unique and sorted, bamm:List is sorted.
```
Comment (atextor) [inline documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc:494]: "apply also _to_". Also check other occurences of this wording in other places

File: documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc (hunk 6 above)
Comment (atextor) [inline documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc:476]: Why is it "\*\*Property\*\*" and not "bamm:Property" in the first column?

File: documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc [hunk 8]
```diff
@@ -439,3 +439,82 @@ characteristicDefinitionsUrl.ifPresent( url -> {
...
+
+| bamm:Property.preferredName   | aas:Property.displayName  |
+
+| bamm:Property.description   | aas:Property.description  |  Note: Also mapped to aas:DataSpecificationIEC61360.definition of the aas:ConceptDescription of this property
+
+| bamm:Property.exampleValue  |aas:Property.value  |
+
+| bamm:Characteristic.dataType  | aas:Property.valueType |
+
+| **Operation** |  Operation  | in/out parameters are not used in BAMM so the mapping only generates input variables and output variables in AAS
```
Comment (atextor) [inline documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc:490]: Why not "bamm:Operation"? Question also applies to several cells below in this column.

File: documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc [hunk 9]
```diff
@@ -439,3 +439,82 @@ characteristicDefinitionsUrl.ifPresent( url -> {
...
+
+| **Code**  | aas:SubmodelElement, aas:ConceptDescription | Similar to plain Characteristic.
+
+| **StructuredValue**  | aas:SubmodelElement, aas:ConceptDescription | The general remarks to Characteristics apply also for StructuredValue. However, AAS has no concpet like deconstruction rule. Thus, the deconstruction rule and the sub properties of the deconstruction entity will be ignored and only the Characteristic is mapped.
+
+| **Enumeration**  | aas:SubmodelElement, aas:ConceptDescription | The general remarks to Characteristics apply also for Enumerations. Additionally, the values of an Enumeration are mapped to a valueList of a DataSpecificationIEC61360.
+
+| **State**  | aas:SubmodelElement, aas:ConceptDescription | Same as Enumeration.
+
+| bamm:MultiLanguageString  | aas:MultiLanguageProperty | if a multilanguage string is used in BAMM it is mapped to the MultiLanguageProperty in AAS.
```
Comment (atextor) [inline documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc:510]: `bamm-c:MultiLanguageText`

File: documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc [hunk 10]
```diff
@@ -439,3 +439,82 @@ characteristicDefinitionsUrl.ifPresent( url -> {
...
+
+| **Collection**  | aas:SubmodelElementCollection, aas:ConceptDescription  | The general remarks to Characteristics apply also for Collection type Characteristics. However, properties referencing Collections are mapped to SubmodelElementCollections. Specific properties of collections are mapped. bamm:Set is unique, bamm:SortedSet is unique and sorted, bamm:List is sorted.
+
+| **Quantifiable**  | aas:SubmodelElement, aas:ConceptDescription | The general remarks to Characteristics apply also for Quantifiable type Characteristics. Quantifiables (also Duration and Measurement) reference a unit, which is added to the ConceptDescription corresponding the mapped Characteristic.
```
Comment (atextor) [inline documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc:496]: Please add `xrefs` for all elements where it makes sense (i.e., at least all Characteristics). For example, instead of `**Quantifiable**`, I'd write:
`xref:bamm-specification:ROOT:characteristics.adoc#quantifiable-characteristic[bamm-c:Quantifiable]`

File: documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc [hunk 11]
```diff
@@ -439,3 +439,82 @@ characteristicDefinitionsUrl.ifPresent( url -> {
...
+| **Enumeration**  | aas:SubmodelElement, aas:ConceptDescription | The general remarks to Characteristics apply also for Enumerations. Additionally, the values of an Enumeration are mapped to a valueList of a DataSpecificationIEC61360.
+
+| **State**  | aas:SubmodelElement, aas:ConceptDescription | Same as Enumeration.
+
+| bamm:MultiLanguageString  | aas:MultiLanguageProperty | if a multilanguage string is used in BAMM it is mapped to the MultiLanguageProperty in AAS.
+|===
+
+
+=== Known Limitations
+The AAS generator implements a base set of features, which are mapped from BAMM to AAS.
```
Comment (atextor) [inline documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc:515]: Is it "AAS generator" or "AAS Generator"? In the introductory paragraph, the other spelling was used.

File: documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc [hunk 12]
```diff
@@ -439,3 +439,82 @@ characteristicDefinitionsUrl.ifPresent( url -> {
...
+However, there are still limitations:
+
+* Predefined entity mapping (FileResource would be mapped to aas:File)
+* BAMM:Either is mapped to aas:SubmodelElementCollection with two entries for left and right side
```
Comment (atextor) [inline documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc:519]: `bamm-c:Either`

File: tools/bamm-cli/pom.xml [hunk 13]
```diff
@@ -85,6 +85,11 @@
          <artifactId>sds-test-aspect-models</artifactId>
          <scope>test</scope>
       </dependency>
+       <dependency>
```
Comment (atextor) [inline tools/bamm-cli/pom.xml:88]: Formatting seems off

File: tools/bamm-cli/src/main/java/io/openmanufacturing/sds/aspect/to/AspectToAasCommand.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+         + "\" and \"" + XML + "\". Default is \"" + XML + "\"." )
+   private String format = XML;
+
+   @CommandLine.ParentCommand
+   private AspectToCommand parentCommand;
+
+   @Override
+   public void run() {
+      final AspectModelAASGenerator generator = new AspectModelAASGenerator();
+      final Aspect aspect = AspectModelLoader.fromVersionedModelUnchecked(loadModelOrFail(parentCommand.parentCommand.getInput()));
```
Comment (atextor) [inline tools/bamm-cli/src/main/java/io/openmanufacturing/sds/aspect/to/AspectToAasCommand.java:53]: The formatting seems off, please run auto-formatter

File: core/sds-aspect-model-aas-generator/src/test/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASGeneratorTest.java [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.EnumSource;
+
+class AspectModelAASGeneratorTest {
```
Comment (atextor) [inline core/sds-aspect-model-aas-generator/src/test/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASGeneratorTest.java:40]: Please also run code formatter and Save Actions (set fields/local variables `final` whenever possible) in this file

File: core/sds-aspect-model-aas-generator/src/test/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASGeneratorTest.java [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+import io.openmanufacturing.sds.metamodel.Aspect;
+import io.openmanufacturing.sds.metamodel.loader.AspectModelLoader;
+import io.openmanufacturing.sds.test.TestAspect;
+import io.openmanufacturing.sds.test.TestResources;
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+import java.io.*;
```
Comment (atextor) [inline core/sds-aspect-model-aas-generator/src/test/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASGeneratorTest.java:29]: Please remove wildcard imports

**Review by atextor — CHANGES_REQUESTED** [review]

File: documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc (hunk 7 above)
Comment (bs-jokri) [reply documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc:494]: 6 oocurences fixed

File: documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc (hunk 6 above)
Comment (bs-jokri) [reply documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc:476]: done

Comment (bs-jokri) [reply documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc:476]: mistake. Fixed

File: documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc (hunk 10 above)
Comment (bs-jokri) [reply documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc:496]: done

File: tools/bamm-cli/src/main/java/io/openmanufacturing/sds/aspect/to/AspectToAasCommand.java (hunk 14 above)
Comment (bs-jokri) [reply tools/bamm-cli/src/main/java/io/openmanufacturing/sds/aspect/to/AspectToAasCommand.java:53]: done

File: core/sds-aspect-model-aas-generator/src/test/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASGeneratorTest.java (hunk 15 above)
Comment (bs-jokri) [reply core/sds-aspect-model-aas-generator/src/test/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASGeneratorTest.java:40]: done

File: core/sds-aspect-model-aas-generator/src/test/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASGeneratorTest.java (hunk 16 above)
Comment (bs-jokri) [reply core/sds-aspect-model-aas-generator/src/test/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASGeneratorTest.java:29]: done

File: core/sds-aspect-model-aas-generator/src/main/java/io/openmanufacturing/sds/aspectmodel/aas/Context.java (hunk 2 above)
Comment (bs-jokri) [reply core/sds-aspect-model-aas-generator/src/main/java/io/openmanufacturing/sds/aspectmodel/aas/Context.java:42]: done

File: core/sds-aspect-model-aas-generator/src/main/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASVisitor.java [hunk 17]
```diff
@@ -0,0 +1,581 @@
...
+   @Override
+   public AssetAdministrationShellEnvironment visitEither( Either either, Context context ) {
+      List<SubmodelElement> submodelElements = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (bs-jokri) [reply core/sds-aspect-model-aas-generator/src/main/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASVisitor.java:480]: done

File: core/sds-aspect-model-aas-generator/src/main/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASVisitor.java (hunk 1 above)
Comment (bs-jokri) [reply core/sds-aspect-model-aas-generator/src/main/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASVisitor.java:477]: done

File: core/sds-aspect-model-aas-generator/README.md [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+with the AAS xml file in it.
+ 
+Should the generator be integrated into custom implementations, the class `AspectModelAASGenerator`
+with its method `io.openmanufacturing.sds.aspectmodel.aas.AspectModelAASGenerator.generateOutput` is the 
+proper entry point. It expects an `io.openmanufacturing.sds.metamodel.Aspect` and returns a 
+`ByteArrayOutputStream`.
+
+
+## Details to the Mapping Concept
+The mapping rules applied by the generator are explained in the [user documentation](documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc)
```
[the remark this answers is not in the document]
Comment (bs-jokri) [reply core/sds-aspect-model-aas-generator/README.md:33]: done

**Comment by bs-jokri** [comment]
@atextor thank you for your review.  I have addressed the raised issues. Would be great if you can have another look and check if they are all addressed as expected.

File: core/sds-aspect-model-aas-generator/src/test/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASGeneratorTest.java [hunk 19]
```diff
@@ -0,0 +1,368 @@
...
+  @Test
+  void testGenerateAasxFromBammAspectWithListAndAdditionalProperty()
+      throws IOException, DeserializationException {
+    final AssetAdministrationShellEnvironment env =
+        getAssetAdministrationShellFromAspect(TestAspect.ASPECT_WITH_LIST_AND_ADDITIONAL_PROPERTY);
+    assertEquals(3, env.getConceptDescriptions().size());
+    assertEquals(1, env.getSubmodels().size());
+    assertEquals(2, env.getSubmodels().get(0).getSubmodelElements().size());
+
+    final Set<String> semanticIds =
```
Comment (atextor) [inline core/sds-aspect-model-aas-generator/src/test/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASGeneratorTest.java:59]: Why not `Set.of(...)` instead of `Stream.of(...).collect(Collectors.toCollection(HashSet::new))`?
(Also in the other similar code parts in this file)

File: documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc [hunk 20]
```diff
@@ -439,3 +439,80 @@ characteristicDefinitionsUrl.ifPresent( url -> {
...
+
+| xref:bamm-specification:ROOT:characteristics.adoc#structured-value-characteristic[**bamm-c:StructuredValue**]  | aas:SubmodelElement, aas:ConceptDescription | The general remarks to Characteristics apply also to StructuredValue. However, AAS has no concpet like deconstruction rule. Thus, the deconstruction rule and the sub properties of the deconstruction entity will be ignored and only the Characteristic is mapped.
+
+| xref:bamm-specification:ROOT:characteristics.adoc#enumeration-characteristic[**bamm-c:Enumeration**]  | aas:SubmodelElement, aas:ConceptDescription | The general remarks to Characteristics apply also to Enumerations. Additionally, the values of an Enumeration are mapped to a valueList of a DataSpecificationIEC61360.
+
+| xref:bamm-specification:ROOT:characteristics.adoc#state-characteristic[**bamm-c:State**]  | aas:SubmodelElement, aas:ConceptDescription | Same as Enumeration.
+
+| bamm-c:MultiLanguageText  | aas:MultiLanguageProperty | if a MultiLanguageText is used in BAMM it is mapped to the MultiLanguageProperty in AAS.
```
Comment (atextor) [inline documentation/developer-guide/modules/tooling-guide/pages/java-aspect-tooling.adoc:508]: This should also be an xref

**Review by atextor — CHANGES_REQUESTED** [review]

File: core/sds-aspect-model-aas-generator/src/main/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASGenerator.java [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+  /**
+   * Generates an AASX archive file for a given Aspect and writes it to a given OutputStream provided by <code>nameMapper<code/>
+   *
+   * @param aspect the Aspect for which an AASX archive shall be generated
+   * @param nameMapper a Name Mapper implementation, which provides an OutputStream for a given filename
+   * @throws IOException in case the generation can not properly be executed
+   */
+  public void generateAASXFile(final Aspect aspect, final Function<String, OutputStream> nameMapper)
+      throws IOException {
+    try (final OutputStream output = nameMapper.apply(aspect.getName())) {
```
[the remark this answers is not in the document]
Comment (bs-jokri) [reply core/sds-aspect-model-aas-generator/src/main/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASGenerator.java:38]: I configured the discussed code styling. The mentioned formatting is now as you describe. I hope it fixes also other points you might have seen.

File: core/sds-aspect-model-aas-generator/src/main/java/io/openmanufacturing/sds/aspectmodel/aas/Context.java [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+import io.openmanufacturing.sds.metamodel.Property;
+
+public class Context {
+
+  AssetAdministrationShellEnvironment environment;
+  Submodel submodel;
+  Property property;
+  SubmodelElement propertyResult;
+
+  public Context(AssetAdministrationShellEnvironment environment, Submodel ofInterest) {
```
[the remark this answers is not in the document]
Comment (bs-jokri) [reply core/sds-aspect-model-aas-generator/src/main/java/io/openmanufacturing/sds/aspectmodel/aas/Context.java:31]: Applied the code style. Should be ok now. However, `final` is not added automatically. I applied it manually wherever appropriate.

File: core/sds-aspect-model-aas-generator/src/main/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASVisitor.java [hunk 23]
```diff
@@ -0,0 +1,643 @@
...
+
+    context.appendToSubModelElements(visitProperties(aspect.getProperties(), context));
+    context.appendToSubModelElements(visitOperations(aspect.getOperations(), context));
+
+    return context.getEnvironment();
+  }
+
+  private List<SubmodelElement> visitOperations(
+      final List<io.openmanufacturing.sds.metamodel.Operation> elements, final Context context) {
+    return Stream.ofAll(elements).map(i -> map(i, context)).collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (bs-jokri) [reply core/sds-aspect-model-aas-generator/src/main/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASVisitor.java:180]: Probably, I found this somewhere else [1] and mimicked It? Yours looks much better. I applied it to all places I found.

[1] io/openmanufacturing/sds/aspectmodel/generator/jsonschema/AspectModelJsonSchemaVisitor.java:221

File: core/sds-aspect-model-aas-generator/src/test/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASGeneratorTest.java (hunk 19 above)
Comment (bs-jokri) [reply core/sds-aspect-model-aas-generator/src/test/java/io/openmanufacturing/sds/aspectmodel/aas/AspectModelAASGeneratorTest.java:59]: thanks for the hint. Looks much cleaner.

File: core/sds-aspect-model-aas-generator/README.md [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+with the AAS xml file in it.
+ 
+Should the generator be integrated into custom implementations, the class `AspectModelAASGenerator`
+with its method `io.openmanufacturing.sds.aspectmodel.aas.AspectModelAASGenerator.generateOutput` is the 
+proper entry point. It expects an `io.openmanufacturing.sds.metamodel.Aspect` and returns a 
+`ByteArrayOutputStream`.
+
+
+## Details of the Mapping Concept
+The mapping rules applied by the generator are explained in the [user documentation](https://openmanufacturingplatform.github.io/sds-documentation/sds-developer-guide/tooling-guide/java-aspect-tooling.html#details-mapping-aas)
```
Comment (atextor) [inline core/sds-aspect-model-aas-generator/README.md:33]: Please add a full stop at the end of the line

File: core/sds-aspect-model-aas-generator/README.md [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+ 
+Should the generator be integrated into custom implementations, the class `AspectModelAASGenerator`
+with its method `io.openmanufacturing.sds.aspectmodel.aas.AspectModelAASGenerator.generateOutput` is the 
+proper entry point. It expects an `io.openmanufacturing.sds.metamodel.Aspect` and returns a 
+`ByteArrayOutputStream`.
+
+
+## Details of the Mapping Concept
+The mapping rules applied by the generator are explained in the [user documentation](https://openmanufacturingplatform.github.io/sds-documentation/sds-developer-guide/tooling-guide/java-aspect-tooling.html#details-mapping-aas) 
+The rules apply to BAMM v1.0.0 [6] and  AAS Specification Part 1 V3.0RC01 [7].
```
Comment (atextor) [inline core/sds-aspect-model-aas-generator/README.md:34]: This is not a request for a change in the PR, but more of a question: As soon as BAMM 2.0.0 is released (2.0.0-M1 already exists as of now), will there be another PR to update the AASX generator?

**Review by atextor — CHANGES_REQUESTED** [review]

File: core/sds-aspect-model-aas-generator/README.md (hunk 25 above)
Comment (bs-jokri) [reply core/sds-aspect-model-aas-generator/README.md:34]: Yes, I think so. And we also have to file a new PR as soon as the new AAS spec version (v3-R2) is available. So I expect several follow-up PRs for that feature in the near future.

**Review by atextor — APPROVED** [review]
